tree/treeplayer/src/TTreeProcessorMT.cxx

```diff
@@ -91,32 +91,141 @@ std::vector<std::vector<Long64_t>> GetFriendEntries(const std::vector<std::pair<
+std::string TTreeProcessorMT::GetTreeName()
+{
+   std::string treeName;
+
+   if (fFileNames.empty())
+      throw std::runtime_error("Empty list of files and no tree name provided");
+
+   ::TDirectory::TContext ctxt(gDirectory);
+   std::unique_ptr<TFile> f(TFile::Open(fFileNames[0].c_str()));
+   TIter next(f->GetListOfKeys());
+   while (TKey *key = (TKey *)next()) {
+      const char *className = key->GetClassName();
+      if (strcmp(className, "TTree") == 0) {
+         treeName = key->GetName();
+         break;
+      }
+   }
+   if (treeName.empty())
+      throw std::runtime_error("Cannot find any tree in file " + fFileNames[0]);
+
+   return treeName;
+}
+
 ////////////////////////////////////////////////////////////////////////
 /// Constructor based on a file name.
 /// \param[in] filename Name of the file containing the tree to process.
 /// \param[in] treename Name of the tree to process. If not provided,
 ///                     the implementation will automatically search for a
 ///                     tree in the file.
-TTreeProcessorMT::TTreeProcessorMT(std::string_view filename, std::string_view treename) : treeView(filename, treename) {}
+TTreeProcessorMT::TTreeProcessorMT(std::string_view filename, std::string_view treename)
+   : fFileNames({std::string(filename)}), fTreeName(treename.empty() ? GetTreeName() : treename) {}
+
+std::vector<std::string> CheckAndConvert(const std::vector<std::string_view> & views)
+{
+   if (views.empty())
+      throw std::runtime_error("The provided list of file names is empty");
+
+   std::vector<std::string> strings;
+   strings.reserve(views.size());
+   for (const auto &v : views)
+      strings.emplace_back(v);
+   return strings;
+}
 
 ////////////////////////////////////////////////////////////////////////
 /// Constructor based on a collection of file names.
 /// \param[in] filenames Collection of the names of the files containing the tree to process.
 /// \param[in] treename Name of the tree to process. If not provided,
 ///                     the implementation will automatically search for a
 ///                     tree in the collection of files.
-TTreeProcessorMT::TTreeProcessorMT(const std::vector<std::string_view> &filenames, std::string_view treename) : treeView(filenames, treename) {}
+TTreeProcessorMT::TTreeProcessorMT(const std::vector<std::string_view> &filenames, std::string_view treename)
+   : fFileNames(CheckAndConvert(filenames)), fTreeName(treename.empty() ? GetTreeName() : treename) {}
 
-////////////////////////////////////////////////////////////////////////
-/// Constructor based on a TTree.
-/// \param[in] tree Tree or chain of files containing the tree to process.
-TTreeProcessorMT::TTreeProcessorMT(TTree &tree) : treeView(tree) {}
+std::vector<std::string> GetFilesFromTree(TTree &tree)
+{
+   std::vector<std::string> filenames;
+
+   const bool isChain = tree.IsA() == TClassRef("TChain");
```

Comment:
TClassRef is intended to be a 'persistent' pointer to a TClass (i.e. it will be refreshed when the 'state' of the underlying TClass changes).  It is a bit of a cpu waste for one time use.  So either do
```
const bool isChain = tree.IsA() == TChain::Class(): // if you can do #include "TChain.h"
```
or 
```
static TClassRef chainClass("TChain");
const bool isChain = tree.IsA() == chainClass;  // if this routine is 'time critical'
```
or
```
const bool isChain = tree.IsA() == TClass::GetClass("TChain"); // The slowest of all options, does a hash search every time.
```

Reply:
Option b is what was there before in the constructor of `TTreeView` (plus a `const` for `chainClass`).

Reply:
so I'll switch to option A in both places

Reply:
no actually `TClassRef` is only used in constructors, so not too time critical, so I went for the more-readable-and-yet-not-too-wasteful option A